packages/neo-one-node-protocol/src/Node.ts

```diff
@@ -206,6 +210,8 @@ export class Node implements INode {
       }
 
       this.requestBlocks();
+    } else if (peer !== undefined && this.mutableHaltOnSync) {
+      this.halt$.next(true);
```

Comment:
Change this to just a callback that you thread through, it'll be a lot more clear than trying to pass it back up through the observable. So just call something here like `this.onSynchronized`

Reply:
Also, I'm not sure this is the right place to do this. What happens if `peer` is `undefined` or `block.index > peer.data.startHeight`? (line 188)

Reply:
Take a look at how we check if the node is "ready" (i.e. fully synchronized) here: https://github.com/neo-one-suite/neo-one/blob/master/packages/neo-one-node-http-rpc/src/middleware/checkReady.ts. Should probably just do that on an interval until it's true.

Reply:
> Also, I'm not sure this is the right place to do this. What happens if `peer` is `undefined` or `block.index > peer.data.startHeight`? (line 188)

My thought process here was:
- we check in both `if`s that peer !== undefined
- this peer, since its defined would be the *best* since we define it as:
```
const peer = this.mutableBestPeer;
```
- if our best peer is lower index than us, then we must have just lost connection to the `true` best peer we requested from?
- if our best peer is equal to us then we're ready.

Reply:
> Take a look at how we check if the node is "ready" (i.e. fully synchronized) here: https://github.com/neo-one-suite/neo-one/blob/master/packages/neo-one-node-http-rpc/src/middleware/checkReady.ts. Should probably just do that on an interval until it's true.

Where would this interval be taking place?

Reply:
> Change this to just a callback that you thread through, it'll be a lot more clear than trying to pass it back up through the observable. So just call something here like `this.onSynchronized`

so instead of passing it up the observable, I would just pass my FullNode *halted* `stop` function: ```
public async stop(halted?: boolean): Promise<void> {
    if (this.mutableSubscription !== undefined) {
      this.mutableSubscription.unsubscribe();
      this.mutableSubscription = undefined;
      await finalize.wait();

      if (halted) {
        const options = await this.options.options$.pipe(take(1)).toPromise();
        const backupOptions = options.backup !== undefined ? options.backup.options : undefined;
        if (backupOptions !== undefined) {
          await this.backup(backupOptions);
        }
      }
    }
  }
```
to the Node itself to call?